### PR TITLE
Neeru auth

### PR DESCRIPTION
**Pull request by neeruram1**
Completed log in functionality with sad path accounted for. All tests passing.

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+class SessionsController < ApplicationController
+  def create
+    @user = User.update_or_create(request.env['omniauth.auth'])
```
Comment (sienna-kopf) [inline app/controllers/sessions_controller.rb:3]: This update or create method is clever!

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,3 @@
+class UsersController < ApplicationController
```
Comment (sienna-kopf) [inline app/controllers/users_controller.rb:1]: Love how clean this controller is. Nice work!

File: app/controllers/welcome_controller.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+class WelcomeController < ApplicationController
+  def index; end
+
+  def failure
```
Comment (sienna-kopf) [inline app/controllers/welcome_controller.rb:4]: Nice addition of a sad path here for the login!

File: app/views/layouts/application.html.erb [hunk 4]
```diff
@@ -10,6 +10,10 @@
   </head>
 
   <body>
+    <% flash.each do |type, message| %>
```
Comment (sienna-kopf) [inline app/views/layouts/application.html.erb:13]: Always smart to put this in the application.html.erb file!! Works so well here!

File: app/views/welcome/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+<h1>WeatherBop</h1>
+
+<%= link_to "log in with spotify", "/auth/spotify" %>
+
+<p>don't have a spotify account? <%= link_to "register here", "https://www.spotify.com/us/" %></p>
```
Comment (sienna-kopf) [inline app/views/welcome/index.html.erb:5]: Nice edge case handling here. This makes it really easy for our user to sign up for a spotify account!!

File: config/initializers/omniauth.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
+require 'omniauth-oauth2'
+
+module OmniAuth
```
Comment (sienna-kopf) [inline config/initializers/omniauth.rb:3]: Love the use of Modules here!

File: config/initializers/omniauth.rb [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+      def callback_url
+        full_host + script_name + callback_path
+      end
+    end
+  end
+end
+
+Rails.application.config.middleware.use OmniAuth::Builder do
+  provider :spotify, ENV['SPOTIFY_CLIENT_ID'], ENV['SPOTIFY_CLIENT_SECRET'], scope: %w(
```
Comment (sienna-kopf) [inline config/initializers/omniauth.rb:33]: I like the scopes you included! As the need arises I'm sure we can work to include more!

File: config/routes.rb [hunk 8]
```diff
@@ -1,3 +1,6 @@
 Rails.application.routes.draw do
-  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+  get '/', to: 'welcome#index'
+  get '/auth/spotify/callback', to: 'sessions#create'
+  get '/auth/failure', to: 'welcome#failure'
+  get '/dashboard', to: 'users#show'
```
Comment (sienna-kopf) [inline config/routes.rb:5]: Smart to send the dashboard to the Users controller show action. This makes sense!

File: spec/features/welcome/index_spec.rb [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+          'display_name' => 'Neeru Ram',
+          'id'           => '12345',
+        },
+        'credentials' => {
+          'token'         => '12345',
+          'refresh_token' => '23456',
+          'email'         => 'neeram85@gmail.com'
+        }
+      }
+    OmniAuth.config.mock_auth[:spotify] = auth_data
```
Comment (sienna-kopf) [inline spec/features/welcome/index_spec.rb:23]: Nice use of mock auth in this test!

File: spec/features/welcome/index_spec.rb [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+      visit '/'
+
+      click_on "log in with spotify"
+
+      expect(current_path).to eq("/dashboard")
+      expect(page).to have_content("welcome, #{neeru.name}")
+  end
+
+  it "has a link to register with spotify" do
```
Comment (sienna-kopf) [inline spec/features/welcome/index_spec.rb:64]: Love this edge case test!

File: spec/features/welcome/index_spec.rb [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    visit '/'
+
+    expect(page).to have_content("don't have a spotify account?")
+    expect(page).to have_link('register here', href: 'https://www.spotify.com/us/')
+  end
+
+  it "redirects to welcome page if authorization fails" do
```
Comment (sienna-kopf) [inline spec/features/welcome/index_spec.rb:71]: Love this sad path test! Very thorough!
